experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts

```diff
@@ -0,0 +1,167 @@
+	}
+
+	public compose(changes: TaggedChange<SharedTreeChange>[]): SharedTreeChange {
+		const newChanges: Mutable<SharedTreeChange["changes"]> = [];
+		const dataChangeRun: TaggedChange<ModularChangeset>[] = [];
+
+		const flushDataChangeRun = (): void => {
+			if (dataChangeRun.length > 0) {
+				newChanges.push({
+					type: "data",
+					innerChange: this.modularChangeFamily.compose(dataChangeRun),
+				});
+				dataChangeRun.length = 0;
+			}
+		};
+
+		for (const topChange of changes) {
+			for (const change of topChange.change.changes) {
+				if (change.type === "schema") {
+					flushDataChangeRun();
+					newChanges.push(change);
+				} else {
+					dataChangeRun.push(mapTaggedChange(topChange, change.innerChange));
+				}
+			}
+		}
+
+		flushDataChangeRun();
+		return { changes: newChanges };
+	}
+
+	public invert(change: TaggedChange<SharedTreeChange>, isRollback: boolean): SharedTreeChange {
+		const invertInnerChange: (
+			innerChange: SharedTreeChange["changes"][number],
+		) => SharedTreeChange["changes"][number] = (innerChange) => {
+			switch (innerChange.type) {
+				case "data":
+					return {
+						type: "data",
+						innerChange: this.modularChangeFamily.invert(
+							mapTaggedChange(change, innerChange.innerChange),
+							isRollback,
+						),
+					};
+				case "schema": {
+					return {
+						type: "schema",
+						innerChange: {
+							schema: {
+								new: innerChange.innerChange.schema.old,
+								old: innerChange.innerChange.schema.new,
+							},
+						},
+					};
+				}
+				default:
+					fail("Unknown SharedTree change type.");
+			}
+		};
+		return {
+			changes: change.change.changes.map(invertInnerChange).reverse(),
+		};
+	}
+
+	public rebase(
+		change: SharedTreeChange,
+		over: TaggedChange<SharedTreeChange>,
+		revisionMetadata: RevisionMetadataSource,
+	): SharedTreeChange {
+		/**
+		 * Any SharedTreeChange (a list of sub-changes) that contains a schema change will cause ANY change that rebases over it to conflict.
+		 * Similarly, any SharedTreeChange containing a schema change will fail to rebase over ANY change.
+		 * Those two combine to mean: no concurrency with schema changes is supported.
+		 * This is fine because it's an open problem. Example: a tree with an A at the root and a schema that allows an A | B at the root will
+		 * become out of schema if a schema range to restrict root types to just A is concurrent with a data change that sets it to a B.
+		 * We don't have an efficient way to detect this document-wide and there are varying opinions on restricting schema changes to prevent this.
+		 * A SharedTreeChange containing a schema change will NOT conflict in a non-concurrency case, as the "meatless sandwich" optimization
+		 * will result in rebase never being called.
+		 */
+		const hasSchemaChange = (c: SharedTreeChange) =>
+			c.changes.some((innerChange) => innerChange.type === "schema");
+		if (
+			change.changes.length === 0 ||
```

Comment:
This may be personal preference, but when `change.changes.length === 0`, I would find it less confusing to return `change`. Especially since there's a coment in the scope below talking about conflict which this isn't.